File: jsrc/kernelprolog/fluents/SxxMachine/SourceLoop.java
package prolog.fluents;

import java.util.ArrayList;

import SxxMachine.Term;
import prolog.terms.Prog;
import prolog.terms.Source;

/**
  An Infinite Source.
  If based on a finite Source, it moves to its the first element after reaching its last element.
  A SourceLoop returns 'no' if the original Source is empty. 
  In case the original Source is infinite, a SourceLoop will 
  return the same elements as the original Source. 
  (In particular, this happens if the original Source is also a Source loop).
*/
public class SourceLoop extends Source {
	private ArrayList v;

	Source s;

	private int i;

	public SourceLoop(Source s, Prog p) {
		super(p);
		this.s = s;
		this.v = new ArrayList();
		this.i = 0;
	}

	private final Term getMemoized() {
		if (null == v || v.size() <= 0)
			return null;
		Term T = (Term) v.get(i);
		i = (i + 1) % v.size();
		return T;
	}

	public Term getElement() {
		Term T = null;
		if (null != s) { // s is alive
			T = s.getElement();
			if (null != T)
				v.add(T);
			else {
				s = null;
			}
		}
		if (null == s)
			T = getMemoized();
		return T;
	}

	public void stop() {
		v = null;
		s = null;
	}
}

File: src/kernelprolog/fluents/SxxMachine/SourceLoop.cpp
using namespace std;

#include "SourceLoop.h"
#include "../../terms/SxxMachine/Prog.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../main/SxxMachine/Builtins.h"

namespace SxxMachine
{

	SourceLoop::SourceLoop(Source *s, Prog *p) : Source(p)
	{
		this->s = s;
		this->v = vector();
		this->i = 0;
	}

	Term *SourceLoop::getMemoized()
	{
		if (nullptr == v || v.size() <= 0)
		{
			return nullptr;
		}
		Term *T = static_cast<Term*>(v[i]);
		i = (i + 1) % v.size();
		return T;
	}

	Term *SourceLoop::getElement()
	{
		Term *T = nullptr;
		if (nullptr != s)
		{ // s is alive
			T = s->getElement();
			if (nullptr != T)
			{
				v.push_back(T);
			}
			else
			{
				s = nullptr;
			}
		}
		if (nullptr == s)
		{
			T = getMemoized();
		}
		return T;
	}

	void SourceLoop::stop()
	{
		v.clear();
		s = nullptr;
	}
}
